### PR TITLE
handle git+ssh with semver

### PR DESCRIPTION
**Pull request by andreineculau**
5 years later I'm still pestering you with git+ssh+semver @zkochan :)

Currently pnpm fails to install `git+ssh://git@example.com:org/repo.git#semver:~1.2.3` (where example.com is an internal git server, not github.com) because it will end up looking for refspec `/~1.2.3` which doesn't match any git tag. We end up there both because of escaping the colon in `semver`, but also because we assume git+ssh URL cannot accept semver range spec.

Trying to install `https://example.com/org/repo.git#semver:~1.2.3` works as expected.

Looking through the code, we had quite a few patches to handle SCP-like urls, because we wanted to use `new URL`, and not the deprecated `url.parse`.

Yarn has this patch https://github.com/yarnpkg/yarn/blob/master/src/util/git.js#L103
but npm https://github.com/npm/git/blob/f60bb15da404679160c4bb6ad1c31a66d5f6ae72/lib/clone.js#L20 was mentioning the shim in https://github.com/npm/hosted-git-info/blob/f03bfbd3022c8f6283a991ff879ed97704ac35fa/lib/parse-url.js#L42

The shim didn't work out, so a similar take as yarn's landed in this PR.

### CODE REVIEW

File: resolving/git-resolver/src/parsePref.ts [hunk 1]
```diff
@@ -151,18 +137,43 @@ function setGitCommittish (committish: string | null) {
...
+    // :password@hostname.com:user/repo
+    // username:password@hostname.com:user/repo
+    // proto://username@hostname.com:user/repo
+    // proto://:password@hostname.com:user/repo
+    // proto://username:password@hostname.com:user/repo
+    // then we replace the last : with a / to create a valid path
+    giturl = giturl.slice(0, lastColonBeforeHash) + '/' + giturl.slice(lastColonBeforeHash + 1)
   }
+
+  if (lastIndexOfBefore(giturl, ':', '#') === -1 && giturl.indexOf('//') === -1) {
```
Comment (andreineculau) [inline resolving/git-resolver/src/parsePref.ts]: follow pnpm's conventions

```suggestion
  if (lastIndexOfBefore(giturl, ':', '#') === -1 && giturl.includes('//')) {
```

**Comment by andreineculau** [comment]
offtopic: it is great to see pnpm can install git dependencies without a package.json 🎉 🚀 pnpm is a generic package manager 😱 Is that true? If so, I L O V E I T 😍 🤩

**Comment by zkochan** [comment]
Could you add some tests?

**Comment by andreineculau** [comment]
I added two tests (SCP-like and regular URL) for semver range. Both are failing on the current version highlighting the two bugs:
* test with SCP-like URL - fails because it doesn't look for ranges
* test with regular URL - fails because it tries to match `/~x.y.z` with semver tags

**Comment by zkochan** [comment]
ok, also add a changset for your changes.

```
pnpm -w changeset
```

**Review by zkochan — APPROVED** [review]

**Pull request merged**